olp-cpp-sdk-dataservice-read/src/StreamLayerClientImpl.cpp

```diff
@@ -280,6 +282,102 @@ client::CancellableFuture<DataResponse> StreamLayerClientImpl::GetData(
                                                       std::move(promise));
 }
 
+client::CancellationToken StreamLayerClientImpl::Poll(
+    PollResponseCallback callback) {
+  auto poll_task = [=](client::CancellationContext context) -> PollResponse {
+    std::string subscription_id;
+    std::string subscription_mode;
+    std::string x_correlation_id;
+    std::shared_ptr<client::OlpClient> client;
+
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (!client_context_) {
+        OLP_SDK_LOG_WARNING_F(kLogTag,
+                              "Poll: unsuccessful, not subscribed");
+
+        return client::ApiError(client::ErrorCode::PreconditionFailed,
+                                "Not subscribed", false);
+      }
+
+      subscription_id = client_context_->subscription_id;
+      subscription_mode = client_context_->subscription_mode;
+      x_correlation_id = client_context_->x_correlation_id;
+      client = client_context_->client;
+    }
+
+    OLP_SDK_LOG_DEBUG_F(kLogTag,
+                        "Poll: started, subscription_id=%s, "
+                        "subscription_mode=%s, x_correlation_id=%s",
+                        subscription_id.c_str(), subscription_mode.c_str(),
+                        x_correlation_id.c_str());
+
+    auto data =
+        StreamApi::ConsumeData(*client, layer_id_, subscription_id,
+                               subscription_mode, context, x_correlation_id);
+
+    if (!data.IsSuccessful()) {
+      OLP_SDK_LOG_WARNING_F(kLogTag, "Poll: consume data unsuccessful, error=%s",
+                            data.GetError().GetMessage().c_str());
+      return data.GetError();
+    }
+
+    auto res = data.MoveResult();
+
+    const auto& messages = res.GetMessages();
+
+    if (!messages.empty()) {
+      // Get offsets for all partitions presented in messages.
+      struct Compare {
+        bool operator()(const model::Message& lhs,
+                        const model::Message& rhs) const {
+          return lhs.GetOffset().GetPartition() <
+                 rhs.GetOffset().GetPartition();
+        }
+      };
+      std::set<model::Message, Compare> stream_offsets(messages.rbegin(),
```

Comment:
This is a copy of the message, including probably the less than 1 MBytes Data. I don't think you want that. Just create a set if necessary from the offsets only.